sbi/utils/posterior_ensemble.py

```diff
@@ -0,0 +1,446 @@
+                up the ensemble.
+
+        Raises:
+            AssertionError if ensemble components have different device variables.
+
+        Returns:
+            A device string, that is the same for all posteriors.
+        """
+        devices = [posterior._device for posterior in posteriors]
+        assert all(
+            device == devices[0] for device in devices
+        ), "Only supported if all posteriors are on the same device."
+        return devices[0]
+
+    @property
+    def weights(self):
+        return self._weights
+
+    @weights.setter
+    def weights(self, weights: List[float] or Tensor):
+        """Set relative weight for each posterior in the ensemble.
+
+        Weights are normalised.
+
+        Args:
+            weights: Assignes weight to each posterior distribution.
+        """
+        if weights is None:
+            self._weights = torch.tensor(
+                [1.0 / self.num_components for _ in range(self.num_components)]
+            )
+        elif type(weights) == Tensor or type(weights) == List:
+            self._weights = torch.tensor(weights) / sum(weights)
+        else:
+            raise TypeError
+
+    def sample(self, sample_shape: Shape = torch.Size(), **kwargs) -> Tensor:
+        r"""Return samples from posterior ensemble.
+
+        All kwargs are passed directly through to `posterior.sample()`.
+
+        Args:
+            sample_shape: Desired shape of samples that are drawn from posterior
+                ensemble. If sample_shape is multidimensional we simply draw
+                `sample_shape.numel()` samples and then reshape into the desired shape.
+
+        Returns:
+            Samples drawn from the ensemble distribution.
+        """
+        num_samples = torch.Size(sample_shape).numel()
+        idxs = torch.multinomial(self._weights, num_samples, replacement=True)
+        samples = []
+        for comp_idx, sample_size in torch.vstack(idxs.unique(return_counts=True)).T:
+            sample_shape_c = torch.Size((sample_size,))
+            samples.append(self.posteriors[comp_idx].sample(sample_shape_c, **kwargs))
+        return torch.vstack(samples).reshape(*sample_shape, -1)
+
+    def log_prob(self, theta: Tensor, individually: bool = False, **kwargs) -> Tensor:
+        r"""Returns the average log-probability of the posterior ensemble
+
+        $\sum_{i}^{N} w_{i} p_i(\theta|x)$.
+
+        All kwargs are passed directly through to `posterior.log_prob()`.
+
+        Args:
+            theta: Parameters $\theta$.
+            individually: If true, returns log weights and log_probs individually.
+
+        Raises:
+            AssertionError if posterior estimators are a mixture of different methods.
+
+        Returns:
+            `(len(θ),)`-shaped average log posterior probability $\log p(\theta|x)$ for
+            θ in the support of the prior, -∞ (corresponding to 0 probability) outside.
+        """
```

Comment:
Here, you have to call the [_x_else_default_x](https://github.com/mackelab/sbi/blob/8657c55b39ec8f673a0eaa0cb2428b7f2ce5b14f/sbi/inference/posteriors/base_posterior.py#L134-L145) method. If `x=None`, default_x will be used.

Reply:
true!